### ISSUE TITLE
Rationalize baseURL and config.yml handling

### ISSUE THREAD

**Issue by geealbers** [gettypubs/quire#19]
Currently, there are an overwhelming **five** config files in the main project directory:

- `config.yml`
- `config.build.yml`
- `config.dev.yml`
- `config.epub.yml`
- `netlify.toml`

`netlify.toml` is obviously specific for a Netlify deployment, but at this stage, I don't think that should be so primary. Down the road it might be a more central part of Quire, but for now, if someone wants to deploy on Netlfy we can document the process of adding this file, and changing other settings. So, this one can go.

Of the remainder, `config.yml` is the main one, the other three YAML files are only to adjust specific `baseURL`, `canonifyURLs`, and `relativeURLs` values for different environments. Most critical is that when building EPUBs with the Quire CLI, the settings need to be:

```
baseURL: "localhost:1313"
canonifyURLs: true
```

So that's what's set in `config.epub.yml`, but I believe there's no reason those same settings couldn't also work for local development. The `config.dev.yml` file currently does nothing, so it seems to me that the `epub` and `dev` settings could be the same and that they could all live in the main `config.yml` file. 

Finally, we have `config.build.yml` and this is where users could enter their specific URL depending on where they're ultimately deploying the site. And so the instructions on the `config.yml` file settings could be essentially, "Leave them alone".

## And what of the URL?

Connected to this is our use of the main publication URL in a number of partials, particularly for site metadata and for the Cite This Page feature. The Quire content model says that a publication has a `url` (defined in `publication.yml`) and we could use that for all the partials: `{{ .Site.Data.publication.url }}` . Or, we could point those partials to `{{ .Site.BaseURL }}` which when building the site for deployment would be the `baseURL` in `config.build.yml`, and so would avoid duplicating the URL in both that file and in `publication.yml`. However, because the URL is output in content and so would be proofread and have an effect on the layout by means of its length you wouldn't want that changing during the development process. So, better to allow the duplication and use `url` with perhaps the `baseURL` being the fallback if `url` isn't set.

**Comment by geealbers** [comment]
@egardner, whenever you have a moment, could you weigh in to make sure this makes sense and I'm not missing anything?

**Comment by workergnome** [comment]
In general, it's often really useful to be able to specify environment-specific configurations.  For instance, in the somewhat near term, I'd like to be able to offer hosted Continuous Integration, where we build a version on the cloud somewhere and run the tests.  We might not be able to specify the explicit port there, and so we'd need a way to configure it.

**Comment by egardner** [comment]
One of the main reasons for the proliferation of these files (which I agree is starting to get confusing, at least for non-developers) was that I wanted to avoid setting `canonifyURLs: true` anywhere I could. Canonicalization means that all relative URLs are transformed into absolute URLs relative to the current `baseURL` value.

[At one point in the past](https://gohugo.io/content-management/urls/#canonicalization) Hugo behaved this way by default, but the authors decided to change it. Canonicalization has pros and cons.  URLs will be more consistent and less likely to break this way, but the trade-off is that you lose some flexibility: doing things like automatically switching between `http` and `https` based on how a page was accessed, or sharing a developing site to others over tools like `ngrok` will no longer be possible.

However, since a "publication" typically means something more formal and official than just another web site, I think that it would be reasonable to say that Quire projects will canonicalize URLs by default in most environments.

Disabling this and using RelativeURLs in the `config.build.yml` environment would still be useful in many cases however – I'm thinking especially of "staging" deployments to somewhere like Netlify or GitHub Pages, where you need the site to be available at some URL other than whatever will be used upon final publication. 

In the past I had some issues trying to override the default `config.yml` settings in `config.build.yml` when canonicalization was already set in the former file; this is part of why I added a dev-specific config file. But it's possible I was just doing something wrong, I was trial-and-error-ing my way through a lot of setting combinations at that point.

## TL;DR

As always, discussions about URL-handling best-practices can go into the weeds fast. 

My recommendation: **if** we can still keep `config.build.yml` and use relative URLs there instead of canonical ones, then we could set `canonifyURLs: true` in most other environments; this means we could probably get rid of `config.dev.yml` and `config.epub.yml`. Having `netlify.toml` is nice in my opinion, but we could always move it to the docs if that is preferred.

Using canonical URLs **everywhere** (including on build) would also be possible – this would allow you to basically ditch all the environment-specific configuration and go back to having a single `config.yml`. However, this is going to introduce some behavior that users will almost certainly find counter-intuitive and will require additional documentation. I'd recommend trying the above approach first.

In either case, I think the updated settings should be thoroughly tested to make sure we find all bugs and edge cases before we update documentation.

**Comment by geealbers** [comment]
To sum up:

- [ ] Remove: `config.dev.yml`, `config.epub.yml`, and `netlify.toml`.
- [ ] Leave  `config.build.yml` as is.
- [ ] Set `config.yml` to:

  ```
  baseURL: "localhost:1313"
  canonifyURLs: true
  ```
- [ ] Update `quire-cli` to account for changes
- [ ] Update guide/documentation accordingly

And I'm going to break my other original question regarding the handling of `{{ .Site.BaseURL }}` versus `{{ .Site.Data.publication.url }}` into a separate issue.

**Comment by fchasen** [comment]
This is a little controversial, but coming from using Rails, I'd suggest going maximalist on the config files but to just organize them in subfolders. 

The thought being that it is easier to look at several very short and specific config files than a longer one. Also including these in a overkill way for the starter lets us specify a way to handle more complex situations upfront and to pass environment variables.

My main confusion is that some of these configs are for **formats** (web | print | epub) and some are for **environments** (dev | remote | github ect...).

I've tried to split the params up into ones that enable or disable certain features (like search) and ones that depend on the environment (like url, logging & analytics) and then a root config with just the parts every project will need to change (title & theme, maybe url) and custom params. 

Attached structure is just a rough concept, as I'm not sure what all of these do yet:
[config-subfolders.zip](https://github.com/gettypubs/quire/files/1783115/config-subfolders.zip)

**Comment by geealbers** [comment]
you all are driving me crazy

**Comment by geealbers** [comment]
So, I think a large portion of the user base for Quire are people like me, who don't want to see or think about multiple config files. So, putting the extras in a folder is a good idea. However, I also think that we need to be able point users like me to **one** obvious location to make any changes they might need/want to make. And I think users should be able to run `hugo` command alone (like then now have to on a PC) and get the same result they would by running `quire build`. So, if we're separating out config files, I think the bulk of the information has to stay in `config.yml` at the root level, and that any information in those other config files has to be really in the weeds and only stuff developers would care about or need access to. @fchasen If after meeting those criteria, there's still need for multiple config files in a separate directory, that's fine. As for the environments vs. formats question, I think we just call them all environments. Technically the way you have them in the sample, the environments ones are all a sub-category of the `web` one in the formats directory.

**Comment by fchasen** [comment]
Simplified the config folder based on the above, and made a PR in cli and starter with a possible setup for the config files (based mostly on Rails setup).

Agreed that most of what is in those folder won't be important for most situations, and my hope with this setup is the web / print / epub config won't really be touched either.

Not completely familiar with the workflow yet, but I think being able to change the environment config quickly will end being useful in a local > preview (private server) > publish (public server) type situation. 

https://github.com/gettypubs/quire-cli/pull/10
https://github.com/gettypubs/quire-starter/pull/9
